smac_planner/src/a_star.cpp

```diff
@@ -365,6 +384,84 @@ bool AStarAlgorithm<NodeT>::isGoal(NodePtr & node)
 }
 
 template<>
+AStarAlgorithm<NodeSE2>::NodePtr AStarAlgorithm<NodeSE2>::getAnalyticPath(const NodePtr & node, const ValidityChecker & checker)
+{
+  if (_motion_model == MotionModel::DUBIN || _motion_model == MotionModel::REEDS_SHEPP)
+  {
+    std::vector<std::pair<NodePtr, Coordinates>> possible_nodes;
+    const NodePtr & goal = getGoal();
+    ompl::base::ScopedState<> from(node->motion_table.state_space), to(node->motion_table.state_space), s(node->motion_table.state_space);
+    const NodeSE2::Coordinates & node_coords = node->pose;
+    const NodeSE2::Coordinates & goal_coords = goal->pose;
+    from[0] = node_coords.x;
+    from[1] = node_coords.y;
+    from[2] = node_coords.theta * node->motion_table.bin_size;
+    to[0] = goal_coords.x;
+    to[1] = goal_coords.y;
+    to[2] = goal_coords.theta * node->motion_table.bin_size;
+
+    float d = node->motion_table.state_space->distance(from(), to());
+    NodePtr prev(node);
+    // A move of sqrt(2) is guaranteed to be in a new cell
+    constexpr float sqrt_2 = std::pow(2., 0.5);
+    unsigned int num_pts = std::floor(d / sqrt_2) + 1;
+    std::vector<double> reals;
+    // Don't generate the first point because we are already there!
+    // And the last point is the goal, so ignore it too!
+    for (unsigned int i = 1; i < num_pts; i++) {
+      node->motion_table.state_space->interpolate(from(), to(), (double)i/num_pts, s());
+      reals = s.reals();
+      float angle = reals[2] / node->motion_table.bin_size;
+      while (angle >= node->motion_table.num_angle_quantization_float) {
+        angle -= node->motion_table.num_angle_quantization_float;
+      }
+      while (angle < 0.0) {
+        angle += node->motion_table.num_angle_quantization_float;
+      }
+      // Turn the pose into a node, and check if it is valid
+      unsigned int index = NodeSE2::getIndex(
+          static_cast<unsigned int>(reals[0]),
+          static_cast<unsigned int>(reals[1]),
+          static_cast<unsigned int>(angle));
+      // Get the node from the graph
+      NodePtr next(nullptr);
+      if (checker(index, next) && next->isNodeValid(_traverse_unknown) && next != prev) {
```

Comment:
I was just testing, and I set the analytic expansion to every 5 nodes for the fun of it, and I see a ton of collisions, I don't think this is doing the collision checking correctly. Even without changing to `5`, you can see occasional paths at the end from the expansions which are in collision with the environment.

Reply:
If you change the frequency of running to something small like 5, you'll see paths that just cut straight through the map indiscriminately from the starting pose.

Reply:
I fixed it with the following logic

```
      NodePtr next(nullptr);
      if (checker(index, next)) {
        Coordinates initial_node_coords = next->pose;
        next->setPose(Coordinates(reals[0], reals[1], angle));
        if (next->isNodeValid(_traverse_unknown) && next != prev) {
          // Set coordinates
          const Coordinates c(reals[0], reals[1], angle);
          possible_nodes.emplace_back(next, c);
          prev = next;
        } else {
          next->setPose(initial_node_coords);
          return NodePtr(nullptr);
        }
      } else {
        // Abort
        return NodePtr(nullptr);
      }
```

Basically, before doing validity checking, you just need to add the continuous state pose for the collision checking to use. Then if you find that you don't use it, set the pose back.

Reply:
This could be further optimized by pre-allocating the new coordinates variable.

Reply:
Thank you for finding this! Good catch!